velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -70,22 +258,63 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
   // Without explicitly setting flush policy, the default memory based flush
   // policy is used.
 
-  auto hiveWriterParameters =
-      std::dynamic_pointer_cast<const HiveWriterParameters>(
-          writeProtocol_->getWriterParameters(
-              insertTableHandle_, connectorQueryCtx_));
-  VELOX_CHECK_NOT_NULL(
-      hiveWriterParameters,
-      "Hive data sink expects write parameters for Hive.");
-  writerParameters_.emplace_back(hiveWriterParameters);
-
-  auto writePath = fs::path(hiveWriterParameters->writeDirectory()) /
-      hiveWriterParameters->writeFileName();
+  auto writePath = fs::path(writerParameters->writeDirectory()) /
+      writerParameters->writeFileName();
   auto sink = dwio::common::DataSink::create(writePath);
   return std::make_unique<Writer>(
       options, std::move(sink), *connectorQueryCtx_->memoryPool());
 }
 
+void HiveDataSink::createWriterForPartition(vector_size_t partitionId) {
+  if (writers_.size() < (partitionId + 1)) {
+    writers_.resize(partitionId + 1);
+    writerParameters_.resize(partitionId + 1);
+  }
+
+  std::optional<std::string> partitionDirectory = partitionChannels_.size() > 0
+      ? std::optional<std::string>(getPartitionString(
+            input_,
+            partitionChannels_,
+            partitionRows_[partitionId]->as<vector_size_t>()[0]))
+      : std::nullopt;
+  auto writerParameters = std::dynamic_pointer_cast<const HiveWriterParameters>(
+      writeProtocol_->getWriterParameters(
+          insertTableHandle_,
+          connectorQueryCtx_,
+          std::make_shared<HiveConnectorWriteInfo>(partitionDirectory)));
+  VELOX_CHECK_NOT_NULL(
+      writerParameters, "Hive data sink expects write parameters for Hive.");
+  writerParameters_[partitionId] = writerParameters;
+
+  writers_[partitionId] = createWriter(writerParameters);
+}
+
+void HiveDataSink::moveWriters(const std::map<uint64_t, uint64_t>& idMap) {
+  uint64_t maxNewIndex = idMap.rbegin()->second;
+
+  if (writers_.size() < maxNewIndex + 1) {
+    writers_.resize(maxNewIndex + 1);
+    writerParameters_.resize(maxNewIndex + 1);
+  }
+  // Rehash monotonically makes the IDs of previous partition values larger. So
+  // moving the existing writers to the high indices starting with the high
+  // end is safe and will not overwrite existing ones.
+  for (vector_size_t i = writers_.size() - 1; i >= 0; i--) {
```

Comment:
nit: Any particular reason not to loop over idMap like you do in the ID generator?

Reply:
You're right. That'll cost less.